### ISSUE TITLE
BroadLink SP Mini Wi-Fi switch and BroadLinkRM Mini 3

### ISSUE THREAD

**Issue by astrojetson8** [lprhodes/homebridge-broadlink-rm#23]
Every time I tried to use learn toggle switch on my iPhone, I would get following error. 
```
Discovered Broadlink RM device at 192.168.1.4 (37:27:04:01:a8:c0)
/usr/local/lib/node_modules/homebridge-broadlink-rm/helpers/learnData.js:59
  device.enterLearning()
         ^

TypeError: device.enterLearning is not a function
    at Object.start (/usr/local/lib/node_modules/homebridge-broadlink-rm/helpers/learnData.js:59:10)
    at LearnIRAccessory.toggleLearning (/usr/local/lib/node_modules/homebridge-broadlink-rm/accessories/learnIR.js:19:17)
    at emitMany (events.js:127:13)
    at emit (events.js:201:7)
    at Characteristic.setValue (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Characteristic.js:155:10)
    at Bridge.<anonymous> (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:740:22)
    at Array.forEach (native)
    at Bridge.Accessory._handleSetCharacteristics (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:685:8)
    at emitMany (events.js:127:13)
    at HAPServer.emit (events.js:201:7)
```

I just noticed this plugin was picking up 2 of my BroadLink SP Mini Wi-Fi switch as well as BroadLinkRM Mini 3. The mac address for all 3 were incorrect but IP address was correct. 

For example:
Discovered Broadlink RM device at 192.168.1.4 (37:27:04:01:a8:c0)

So, I turned off both of BroadLink SP Mini Wi-Fi switch, this plugin works. Looks like I found the root cause of the above conflict. It might be helpful for others. Let me know if you want me troubleshoot anything.

I was also wondering, can you control BroadLink SP Mini Wi-Fi with your plugin?

**Comment by lprhodes** [comment]
Hey - thanks for reporting this. I don't have any other devices but the Broadlink RM to it's useful to get these kind of reports.

I've fixed the issue you were seeing in 1.0.4 so you can have all the devices on the network.

If you were able to get the decrypted hex codes for the BroadLink SP Mini Wi-Fi with as packet sniffer  like Wireshark (for IR/RF we normally get through the Learn IR function) then I don't see why you wouldn't be able to use them.

You could tweak broadlinkjs to pass the encrypted hex code through to get the decrypted one back.

**Comment by astrojetson8** [comment]
Thanks! It's all good now. 

I wouldn't have a clue where to begin with Wireshark...

**Comment by astrojetson8** [comment]
I just got this error when I asked to turn on a device. I can't seem to replicate this issue.

[3/30/2017, 4:26:09 PM] [Broadlink RM] setSwitchState: true
/usr/local/lib/node_modules/homebridge-broadlink-rm/helpers/sendData.js:31
  device.sendData(packet);
         ^

TypeError: device.sendData is not a function
    at module.exports (/usr/local/lib/node_modules/homebridge-broadlink-rm/helpers/sendData.js:31:10)
    at SwitchAccessory.setSwitchState (/usr/local/lib/node_modules/homebridge-broadlink-rm/accessories/switch.js:21:5)
    at emitMany (events.js:127:13)
    at emit (events.js:201:7)
    at Characteristic.setValue (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Characteristic.js:155:10)
    at Bridge.<anonymous> (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:740:22)
    at Array.forEach (native)
    at Bridge.Accessory._handleSetCharacteristics (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:685:8)
    at emitMany (events.js:127:13)
    at HAPServer.emit (events.js:201:7)

**Comment by lprhodes** [comment]
I assume that's with your SP switches? Looking at broadlinkjs it appears SP's works differently.

I can add support for them but it won't be until next week.

**Comment by lprhodes** [comment]
Also, make sure you're setting the host value for each accessory given that you have multiple broadlink devices.

**Comment by astrojetson8** [comment]
Sounds good! That would be great!
 
Personally, I'm looking to ditch SP plugs as I got Xiaomi one's now :) Just waiting for new cables as Xiaomi requires earth for AU cables whereas Broadlink doesn't.

But it I might be helpful for others?

You've a great plugin here. I hope apple will update Home to accomodate better UI for remote control.

**Comment by lprhodes** [comment]
You don’t like grounding? :)

It’d at least be good for me to check whether the device is capable of receiving RF/IR rather than crashing.

We should hopefully find out about new UI in June with WWDC.

**Comment by astrojetson8** [comment]
as far as i could find, there's no 3 pin for 8 figure cable so I end up getting US 8 figure cable.

**Comment by lprhodes** [comment]
I've added a check to 1.0.4 so that it doesn't crash.

**Comment by lprhodes** [comment]
@astrojetson8 MAC address has been fixed in 1.1.19 (check out the release notes if it doesn't work)

Hopefully it's correct for all three of your devices, otherwise it may be different for each type of device.
